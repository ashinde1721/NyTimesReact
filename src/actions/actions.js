export const SEARCH_TEXT = "SEARCH_TEXT";
export const SAVE_ARTICLES = "SAVE_ARTICLES";
export const FETCH_ARTICLES = "FETCH_ARTICLES";

export const saveArticles = (articles) => {
    return {
        type: SAVE_ARTICLES,
        payload: {
            articlesList: articles
        }
    };
};

export const fetchArticles = () => {
    return {
        type: FETCH_ARTICLES
    }
}
