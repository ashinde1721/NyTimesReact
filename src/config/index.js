export const API_URL = 'https://api.nytimes.com/svc/mostpopular/v2/mostviewed/all-sections';
export const API_KEY = 'qG808Xn6G5zAb6Wm0e6hS52JQjS6L2Er';
export const PERIOD = 7;