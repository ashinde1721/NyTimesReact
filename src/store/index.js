import { createStore, applyMiddleware } from "redux";
import rootReducer from "./reducer";
import createSagaMiddleware from 'redux-saga';

import {getArticleSaga} from  '../sagas/articlesSaga'

const sagaMiddleware = createSagaMiddleware();

const store = createStore(rootReducer, applyMiddleware(sagaMiddleware));
sagaMiddleware.run(getArticleSaga);

export default store;