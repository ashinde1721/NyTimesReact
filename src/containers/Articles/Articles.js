import React, { Component } from 'react';
import Article from '../../components/Article/Article';
import { connect } from 'react-redux';
import * as actionCreators from '../../actions/actions'

class Articles extends Component {
    constructor(props) {
        super(props);
        this.state = {
            period: 7,
            err: false
        };
    }

    componentDidMount() {
        this._isMounted = true;
        this.props.fetchArticles();
      }
    
      componentWillUnmount() {
        this._isMounted = false;
      }

    goToDetails = (url) => () => {
        this.props.history.push(`/details?url=${url}`);
    }

    render() {
        let articles = [...this.props.articlesList];
        if (this.props.search) {
            articles = this.props.articlesList.filter(article => {
                if (article.title.toLowerCase().includes(this.props.search.toLowerCase())) {
                    return article;
                }
                return null;
            })
        }
        return (
            <div className="row">
                <div className="col-md-12">
                    {
                        articles.map((article, index) => {
                            return (
                                <Article key={article.id} article={article} goToDetails={this.goToDetails} />
                            );
                        })
                    }
                </div>
            </div>
        );
    }
}

const mapStateToProps = state => {
    return {
        search: state.search,
        articlesList: state.articlesList
    }
}

const mapDispatchToProps = dispatch => {
    return {
        fetchArticles: () => dispatch(actionCreators.fetchArticles())
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(Articles);
