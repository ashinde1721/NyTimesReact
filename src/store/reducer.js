import * as actions from '../actions/actions';

const initialState = {
    search: '',
    articlesList: []
};

const reducer = (state = initialState, action) => {
    const { type, payload = {} } = action;

    switch (type) {
        case actions.SEARCH_TEXT:
            return { ...state, search: payload.search };

        case actions.SAVE_ARTICLES:
            return {
                ...state,
                articlesList: payload.articlesList
            }
        default:
            return state;
    }
};

export default reducer;
