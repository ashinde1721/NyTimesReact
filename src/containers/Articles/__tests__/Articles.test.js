import React from 'react';
import { Provider } from 'react-redux';
import axios from 'axios';
import { cleanup, render, waitForElement } from 'react-testing-library';

import store from '../../../store';
import Articles from '../Articles';
import { API_URL, API_KEY } from '../../../config/index'
import { SEARCH_TEXT } from '../../../actions/actions';


describe('Articles Component', () => {
  let renderedComponent;
  const fetchArticles = async () => {
    const period = 7;
    const url = `${API_URL}/${period}.json?api-key=${API_KEY}`;
    const { data: { results = [] } } = await axios.get(url);
    return results;
  };
  
  beforeEach(() => {
    renderedComponent = render(
      <Provider store={store}>
        <Articles />
      </Provider>
    );
  });

  afterEach(() => cleanup());

  it('renders component', () => {
    expect(renderedComponent.container).toMatchSnapshot();
  });

  it('should render all NY-Times-Articles', async () => {
    const { getAllByTestId } = renderedComponent;

    const results = await fetchArticles();
    await waitForElement(() => getAllByTestId('nytimes-card'));
    expect(getAllByTestId('nytimes-card').length).toEqual(results.length);
  });

  it('should render only matched searchText NY-Times-Articles', async () => {
    const searchText = 'The';
    store.dispatch({
      type: SEARCH_TEXT,
      payload: {
        search: searchText,
    }
    });
    const { getAllByTestId } = renderedComponent;

    const results = await fetchArticles();
    const matchedResults = results.filter(({ title }) => {
      if (title.toLowerCase().includes(searchText.toLowerCase())) {
        return true;
      }
      return false;
    });
    await waitForElement(() => getAllByTestId('nytimes-card'));
    expect(getAllByTestId('nytimes-card').length).toEqual(matchedResults.length);
  });
});
