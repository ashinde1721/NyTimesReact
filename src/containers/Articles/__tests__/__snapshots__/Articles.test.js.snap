// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Articles Component renders component 1`] = `
<div>
  <div
    class="row"
  >
    <div
      class="col-md-12"
    />
  </div>
</div>
`;
